### PR TITLE
Fix bug 1680536: Use pip-compile for Python dependencies

### PR DESCRIPTION
**Pull request by PFischbeck**
All direct dependencies are now in `requirements/*.in` files, and the `requirements/*.txt` files are generated with pip-compile. We provide a script for that, since these files should always be updated all at once.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -237,7 +240,7 @@ a virtualenv to build docs, do this:
     $ cd docs/
     $ virtualenv venv
     $ source venv/bin/activate
-    $ pip install --require-hashes --use-deprecated=legacy-resolver -r requirements/default.txt
+    $ pip install --require-hashes -r requirements.txt
```
Comment (mathjazz) [inline CONTRIBUTING.rst:243]: Nice catch with the file name/path :)

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -348,7 +351,7 @@ steps, as they don't affect your setup if nothing has changed:
    git pull origin master
 
    # Install new dependencies or update existing ones.
-   pip2 install -U --force --require-hashes --use-deprecated=legacy-resolver -r requirements/default.txt
+   pip install -U --force --require-hashes -r requirements/default.txt
```
Comment (mathjazz) [inline CONTRIBUTING.rst:354]: pip2 :)

File: docs/dev/setup-virtualenv.rst [hunk 3]
```diff
@@ -68,8 +68,7 @@ Installation
 
    .. code-block:: bash
 
-      pip install --require-hashes --use-deprecated=legacy-resolver -r requirements/dev.txt
-      pip install --require-hashes --use-deprecated=legacy-resolver -r requirements/test.txt
+      pip install --require-hashes-r requirements/dev.txt -r requirements/test.txt
```
Comment (mathjazz) [inline docs/dev/setup-virtualenv.rst:71]: Mind the missing space in front of the first `-r`.

File: requirements/compile_all.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# This compiles all requirements files with pip-compile.
+# You should always use this script, because dev.txt and test.txt depend on default.txt.
+
+export CUSTOM_COMPILE_COMMAND="./requirements/compile_all.sh"
+
+pip-compile --generate-hashes requirements/default.in
+pip-compile --generate-hashes requirements/dev.in
+pip-compile --generate-hashes requirements/test.in
```
Comment (mathjazz) [inline requirements/compile_all.sh:10]: Please add the newline at the end of the file.

**Review by mathjazz — COMMENTED** [review]
Very well executed and documented! I really like this new structure.

Just a couple of minor corrections.

**Review by mathjazz — APPROVED** [review]
Thanks!

Note that we usually submit further changes to the PR as separate commits. When merging into master, we squash all commits into one anyways.

**Comment by PFischbeck** [comment]
Okay, I see. Thanks for reviewing!

**Pull request merged**